### PR TITLE
[Spec_Decode] performance tuning with scheduler overlap compatiblity

### PR DESCRIPTION
**Pull request by pengchengneo**
image

## 4k input / 1k output benchserving logs and results

Raw logs/results bundle: https://gist.github.com/pengchengneo/328518a93ecd11366e5bea09f605d189

The gist contains:
- rank0 raw server logs for spec and no-spec
- benchserving `run.log` and `result.jsonl` for bsz 4/8/16/32/64
- parsed TTFT/ITL CSV
- parsed rank0 steady decode gen-throughput CSV

## Spec server launch command, rank0 shown

```bash
python3 -m sgl_jax.launch_server \
  --model-path /data/pc \
  --trust-remote-code \
  --tp-size 16 \
  --dp-size 4 \
  --ep-size 16 \
  --moe-backend epmoe \
  --nnodes 4 \
  --node-rank 0 \
  --dist-init-addr 10.16.2.6:30042 \
  --host 0.0.0.0 \
  --port 30271 \
  --page-size 256 \
  --context-length 262144 \
  --disable-radix-cache \
  --chunked-prefill-size 2048 \
  --dtype bfloat16 \
  --mem-fraction-static 0.84 \
  --swa-full-tokens-ratio 0.2 \
  --skip-server-warmup \
  --log-level info \
  --decode-log-interval 1 \
  --max-running-requests 128 \
  --dp-schedule-policy round_robin \
  --max-seq-len 4096 \
  --precompile-bs-paddings 1 4 8 16 32 64 128 \
  --precompile-token-paddings 4096 \
  --speculative-algorithm NEXTN \
  --speculative-num-steps 3 \
  --speculative-num-draft-tokens 4 \
  --speculative-eagle-topk 1 \
  --speculative-accept-threshold-single 1.0 \
  --speculative-accept-threshold-acc 1.0
```

## No-spec server launch command, rank0 shown

```bash
python3 -m sgl_jax.launch_server \
  --model-path /data/pc \
  --trust-remote-code \
  --tp-size 16 \
  --dp-size 4 \
  --ep-size 16 \
  --moe-backend epmoe \
  --nnodes 4 \
  --node-rank 0 \
  --dist-init-addr 10.16.2.6:30042 \
  --host 0.0.0.0 \
  --port 30271 \
  --page-size 256 \
  --context-length 262144 \
  --disable-radix-cache \
  --chunked-prefill-size 2048 \
  --dtype bfloat16 \
  --mem-fraction-static 0.84 \
  --swa-full-tokens-ratio 0.2 \
  --skip-server-warmup \
  --log-level info \
  --decode-log-interval 1 \
  --max-running-requests 128 \
  --dp-schedule-policy round_robin \
  --max-seq-len 4096 \
  --precompile-bs-paddings 1 4 8 16 32 64 128 \
  --precompile-token-paddings 4096
```

For ranks 1/2/3, only `--node-rank` and pod/log filename differ.

## Benchserving command template

```bash
for bs in 4 8 16 32 64; do
  num_prompts=$((bs * 3))
  python3 -m sgl_jax.bench_serving \
    --backend sgl-jax \
    --base-url http://127.0.0.1:30271 \
    --model /data/pc \
    --dataset-name random \
    --random-input-len 4096 \
    --random-output-len 1024 \
    --random-range-ratio 1 \
    --max-concurrency "$bs" \
    --num-prompts "$num_prompts" \
    --warmup-requests 0 \
    --output-file "$out/bs_${bs}/result.jsonl" \
    2>&1 | tee "$out/bs_${bs}/run.log"
done
```

## Benchserving results

| bsz | spec out tok/s | no-spec out tok/s | speedup | spec TTFT ms | no-spec TTFT ms | spec ITL ms | no-spec ITL ms |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 4 | 457.90 | 326.17 | 1.404x | 1865.45 | 2091.70 | 6.42 | 10.23 |
| 8 | 711.06 | 519.27 | 1.369x | 2173.81 | 2668.22 | 8.77 | 12.81 |
| 16 | 817.10 | 761.33 | 1.073x | 2691.87 | 3832.60 | 16.21 | 17.28 |
| 32 | 1168.70 | 1103.83 | 1.059x | 4072.27 | 6124.56 | 22.87 | 23.02 |
| 64 | 1162.40 | 1418.37 | 0.820x | 6989.83 | 10875.96 | 46.98 | 34.52 |

## Rank0 server-log steady decode gen throughput

| bsz | spec steady decode gen tok/s | no-spec steady decode gen tok/s | speedup |
| --- | --- | --- | --- |
| 4 | 1024.64 | 402.51 | 2.546x |
| 8 | 1681.14 | 669.43 | 2.511x |
| 16 | 2503.09 | 1046.65 | 2.392x |
| 32 | 4076.77 | 1711.67 | 2.382x |
| 64 | 5025.27 | 2629.48 | 1.911x |

Notes:
- TTFT/ITL are benchserving streaming-client metrics. TTFT includes scheduling and queueing under load, not only first-token compute.
- ITL is inter-stream-chunk latency for `sgl-jax`; it is not retokenized by accepted-token count for speculative decoding in this backend.

## Throughput comparison plot

Spec vs no-spec steady decode throughput comparison

### CODE REVIEW

**Comment by pengchengneo** [comment]
## 4k input / 1k output benchserving logs and results

Raw logs/results bundle: https://gist.github.com/pengchengneo/328518a93ecd11366e5bea09f605d189

The gist contains:
- rank0 raw server logs for spec and no-spec
- benchserving `run.log` and `result.jsonl` for bsz 4/8/16/32/64
- parsed TTFT/ITL CSV
- parsed rank0 steady decode gen-throughput CSV



```bash
python3 -m sgl_jax.launch_server \
  --model-path /data/pc \
  --trust-remote-code \
  --tp-size 16 \
  --dp-size 4 \
  --ep-size 16 \
  --moe-backend epmoe \
  --nnodes 4 \
  --node-rank 0 \
  --dist-init-addr 10.16.2.6:30042 \
  --host 0.0.0.0 \
  --port 30271 \
  --page-size 256 \
  --context-length 262144 \
  --disable-radix-cache \
  --chunked-prefill-size 2048 \
  --dtype bfloat16 \
  --mem-fraction-static 0.84 \
  --swa-full-tokens-ratio 0.2 \
  --skip-server-warmup \
  --log-level info \
  --decode-log-interval 1 \
  --max-running-requests 128 \
  --dp-schedule-policy round_robin \
  --max-seq-len 4096 \
  --precompile-bs-paddings 1 4 8 16 32 64 128 \
  --precompile-token-paddings 4096 \
  --speculative-algorithm NEXTN \
  --speculative-num-steps 3 \
  --speculative-num-draft-tokens 4 \
  --speculative-eagle-topk 1 \
  --speculative-accept-threshold-single 1.0 \
  --speculative-accept-threshold-acc 1.0
```

## No-spec server launch command, rank0 shown

```bash
python3 -m sgl_jax.launch_server \
  --model-path /data/pc \
  --trust-remote-code \
  --tp-size 16 \
  --dp-size 4 \
  --ep-size 16 \
  --moe-backend epmoe \
  --nnodes 4 \
  --node-rank 0 \
  --dist-init-addr 10.16.2.6:30042 \
  --host 0.0.0.0 \
  --port 30271 \
  --page-size 256 \
  --context-length 262144 \
  --disable-radix-cache \
  --chunked-prefill-size 2048 \
  --dtype bfloat16 \
  --mem-fraction-static 0.84 \
  --swa-full-tokens-ratio 0.2 \
  --skip-server-warmup \
  --log-level info \
  --decode-log-interval 1 \
  --max-running-requests 128 \
  --dp-schedule-policy round_robin \
  --max-seq-len 4096 \
  --precompile-bs-paddings 1 4 8 16 32 64 128 \
  --precompile-token-paddings 4096
```

For ranks 1/2/3, only `--node-rank` and pod/log filename differ.

## Benchserving command template

```bash
for bs in 4 8 16 32 64; do
  num_prompts=$((bs * 3))
  python3 -m sgl_jax.bench_serving \
    --backend sgl-jax \
    --base-url http://127.0.0.1:30271 \
    --model /data/pc \
    --dataset-name random \
    --random-input-len 4096 \
    --random-output-len 1024 \
    --random-range-ratio 1 \
    --max-concurrency "$bs" \
    --num-prompts "$num_prompts" \
    --warmup-requests 0 \
    --output-file "$out/bs_${bs}/result.jsonl" \
    2>&1 | tee "$out/bs_${bs}/run.log"
done
```

## Benchserving results

| bsz | spec out tok/s | no-spec out tok/s | speedup | spec TTFT ms | no-spec TTFT ms | spec ITL ms | no-spec ITL ms |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 4 | 457.90 | 326.17 | 1.404x | 1865.45 | 2091.70 | 6.42 | 10.23 |
| 8 | 711.06 | 519.27 | 1.369x | 2173.81 | 2668.22 | 8.77 | 12.81 |
| 16 | 817.10 | 761.33 | 1.073x | 2691.87 | 3832.60 | 16.21 | 17.28 |
| 32 | 1168.70 | 1103.83 | 1.059x | 4072.27 | 6124.56 | 22.87 | 23.02 |
| 64 | 1162.40 | 1418.37 | 0.820x | 6989.83 | 10875.96 | 46.98 | 34.52 |

## Rank0 server-log steady decode gen throughput

| bsz | spec steady decode gen tok/s | no-spec steady decode gen tok/s | speedup |
| --- | --- | --- | --- |
| 4 | 1024.64 | 402.51 | 2.546x |
| 8 | 1681.14 | 669.43 | 2.511x |
| 16 | 2503.09 | 1046.65 | 2.392x |
| 32 | 4076.77 | 1711.67 | 2.382x |
| 64 | 5025.27 | 2629.48 | 1.911x |

Notes:
- TTFT/ITL are benchserving streaming-client metrics. TTFT includes scheduling and queueing under load, not only first-token compute.
- ITL is inter-stream-chunk latency for `sgl-jax`; it is not retokenized by accepted-token count for speculative decoding in this backend.
